src/convert.jl

```diff
@@ -0,0 +1,228 @@
+"""
+function get_nextrefname(record::SAM.Record)
+    !SAM.hasnextrefname(record) && return "*"
+    SAM.nextrefname(record) == "=" ? SAM.refname(record) : SAM.nextrefname(record)
+end
+
+function get_refname(record::SAM.Record)
+    !SAM.hasrefname(record) && return "*"
+    SAM.refname(record)
+end
+
+function set_fixed_length_fields!(bam_record::BAM.Record, record::SAM.Record, header)
+
+    cigar_ops = ('M','I','D','N','S','H','P','=','X')
+    n_cigar_op = sum(x ∈ cigar_ops for x in SAM.cigar(record); init = 0)
+
+    if n_cigar_op > typemax(UInt16) 
+        error("Cigar contains more than $(typemax(UInt16)) operations. This is currently unsupported. See section 4.2.2 of SAM/BAM specifications.")
+    end
+
+    bin = reg2bin(SAM.position(record)-1, SAM.alignlength(record))
+    
+    bam_record.refid = refname_to_refid(get_refname(record), header)
+    bam_record.pos = SAM.position(record)-1
+    bam_record.l_read_name = SAM.tempname(record) |> x -> length(x)+1 # +1 for NULL
+    bam_record.mapq = SAM.mappingquality(record)
+    bam_record.bin = bin
+    bam_record.n_cigar_op = n_cigar_op
+    bam_record.flags = SAM.flags(record)
+    bam_record.l_seq = SAM.hassequence(record) ? SAM.seqlength(record) : 0
+    bam_record.next_refid = refname_to_refid(get_nextrefname(record), header)
+    bam_record.next_pos = SAM.nextposition(record)-1
+    bam_record.tlen = SAM.templength(record)
+    bam_record
+end
+
+"""
+While all single (i.e., non-array) integer types are stored in SAM as 'i', in BAM any of 'cCsSiI' 
+may be used together with the correspondingly-sized binary integer value, chosen according to the field value's magnitude
+"""
+global const type_to_aux_int_fields = Dict(
+    Int8 => UInt8('c'),
+    UInt8 => UInt8('C'),
+    Int16 => UInt8('s'),
+    UInt16 => UInt8('S'),
+    Int32 => UInt8('i'),
+    UInt32 => UInt8('I'),
+)
+
+function find_smallest_int_type(x)
+    candidates = x < 0 ? (Int8, Int16, Int32) : (UInt8, UInt16, UInt32)
+    idx = findfirst(T -> typemin(T) <= x <= typemax(T) , candidates)
+    isnothing(idx) && error("Coulnd't find Int type for value $(x)")
+    T = candidates[idx]
+
+    T, type_to_aux_int_fields[T]
+end
+
+function encode_aux_field(record::SAM.Record, field::UnitRange{Int64})
+    typ = record.data[first(field)+3]
+    key = record.data[first(field):first(field)+1]
+
+    if typ == UInt8('i')
+        val = record.data[first(field)+5:last(field)] |> String
+        val = parse(Int32, val)
+        T, typ = find_smallest_int_type(val)
+        val = reinterpret(UInt8, [T(val)])
+
+        return vcat(key, typ, val)
+    end
+    if typ == UInt8('Z')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val, 0x00)
+    end
+    if typ == UInt8('A')
+        val = record.data[first(field)+5:last(field)]
+        return vcat(key, typ, val)
+    end
+
+    error("Unsupported tag type $(Char(typ))")
```

Comment:
```suggestion
    else
        error("Unsupported tag type $(Char(typ))")
    end
```
Elseif is more efficient (sometimes)